File: extra-code/stage2_code/cpp_action_tutorial/src/count_until_client.cpp
// ============================================================
// Action 通信示例:CountUntil 客户端(Action Client)
// 演示:发送 goal(数到 5,每 0.5 秒一步),通过 feedback 回调实时接收进度,
// result 回调拿到最终结果后退出
// 运行:先启动 count_until_server,再 ros2 run cpp_action_tutorial count_until_client
// ============================================================

#include <memory>
#include <chrono>

#include "count_until_interfaces/action/count_until.hpp"
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"

using CountUntil = count_until_interfaces::action::CountUntil;
// 客户端侧的 GoalHandle,代表"已发出的那个 goal",可用于取消等操作
using GoalHandleCountUntil = rclcpp_action::ClientGoalHandle<CountUntil>;
using namespace std::chrono_literals;

class CountUntilClient : public rclcpp::Node
{
public:
  CountUntilClient() : Node("count_until_client")
  {
    // Action Client 也来自 rclcpp_action,名字必须与服务端的 "count_until" 一致
    client_ = rclcpp_action::create_client<CountUntil>(this, "count_until");
  }

  void sendGoal()
  {
    // 发 goal 前先确认服务端在线,最多等 5 秒
    RCLCPP_INFO(get_logger(), "Waiting for action server...");
    if (!client_->wait_for_action_server(5s)) {
      RCLCPP_ERROR(get_logger(), "Action server not available");
      rclcpp::shutdown();
      return;
    }

    // 填写 goal:数到 5,每步间隔 0.5 秒
    CountUntil::Goal goal;
    goal.target_number = 5;
    goal.period = 0.5;

    // SendGoalOptions 用来注册客户端侧的回调,这里用 lambda(匿名函数)而不是 std::bind,
    // [this] 捕获当前对象以便在 lambda 里调用 get_logger() 等成员
    rclcpp_action::Client<CountUntil>::SendGoalOptions options;
    // feedback 回调:服务端每 publish_feedback 一次就触发一次,实时显示进度
    options.feedback_callback = [this](
      GoalHandleCountUntil::SharedPtr,
      const std::shared_ptr<const CountUntil::Feedback> feedback) {
        RCLCPP_INFO(get_logger(), "Current number: %ld", feedback->current_number);
      };
    // result 回调:goal 结束(成功/取消/中止)时触发一次,拿到最终结果后关闭节点
    options.result_callback = [this](const GoalHandleCountUntil::WrappedResult & result) {
        if (result.result) {
          RCLCPP_INFO(get_logger(), "Reached number: %ld", result.result->reached_number);
        } else {
          RCLCPP_ERROR(get_logger(), "Goal failed without result");
        }
        rclcpp::shutdown();
      };

    // 异步发送 goal,立即返回;后续 feedback/result 都由 spin 的事件循环调用
    client_->async_send_goal(goal, options);
  }

private:
  rclcpp_action::Client<CountUntil>::SharedPtr client_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<CountUntilClient>();
  node->sendGoal();
  // spin 保持节点运行以接收 feedback 和 result;result 回调里的 shutdown 会让 spin 退出
  rclcpp::spin(node);
  return 0;
}

File: extra-code/stage2_code/py_action_tutorial/py_action_tutorial/count_until_client.py
# count_until_client.py —— Action 客户端入门示例
# 演示客户端的完整流程（全部基于异步回调，不阻塞主线程）：
#   发送 Goal -> 服务器答复接受/拒绝 -> 接收 Feedback -> 接收 Result
# 运行（需先启动 count_until_server）：
#   ros2 run py_action_tutorial count_until_client
import rclpy
from action_msgs.msg import GoalStatus
from count_until_interfaces.action import CountUntil
from rclpy.action import ActionClient
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node


class CountUntilClient(Node):
    def __init__(self):
        super().__init__("count_until_client")
        # ActionClient 三要素：所属节点、Action 类型、Action 名称
        # （名称必须与服务器端一致才能连上）
        self.client_ = ActionClient(self, CountUntil, "count_until")

    def send_goal(self, target_number=5, period=0.5):
        # 发送前先等服务器上线，避免 Goal 石沉大海
        self.get_logger().info("Waiting for action server...")
        if not self.client_.wait_for_server(timeout_sec=5.0):
            self.get_logger().error("Action server not available")
            rclpy.shutdown()
            return
        # 填写 Goal 消息并异步发送；send_goal_async 立刻返回 future，
        # 同时注册 feedback_callback 用于接收服务器的进度反馈
        goal = CountUntil.Goal()
        goal.target_number = target_number
        goal.period = period
        future = self.client_.send_goal_async(goal, feedback_callback=self.feedback_callback)
        future.add_done_callback(self.goal_response_callback)

    def feedback_callback(self, feedback_msg):
        # 服务器每次 publish_feedback 都会触发这里
        self.get_logger().info(f"Current number: {feedback_msg.feedback.current_number}")

    def goal_response_callback(self, future):
        # 第一步回应：服务器只告诉我们 Goal 被接受还是拒绝，
        # 被接受后还要再异步请求最终的 Result
        try:
            goal_handle = future.result()
        except Exception as exc:
            self.get_logger().error(f"Goal request failed: {exc}")
            rclpy.shutdown()
            return
        if not goal_handle.accepted:
            self.get_logger().warn("Goal rejected")
            rclpy.shutdown()
            return
        goal_handle.get_result_async().add_done_callback(self.result_callback)

    def result_callback(self, future):
        # 最终 Result 到达：除数据外还带有状态码
        # （SUCCEEDED / CANCELED / ABORTED）
        try:
            wrapped_result = future.result()
        except Exception as exc:
            self.get_logger().error(f"Result request failed: {exc}")
            rclpy.shutdown()
            return
        result = wrapped_result.result
        status_name = {
            GoalStatus.STATUS_SUCCEEDED: "SUCCEEDED",
            GoalStatus.STATUS_CANCELED: "CANCELED",
            GoalStatus.STATUS_ABORTED: "ABORTED",
        }.get(wrapped_result.status, str(wrapped_result.status))
        self.get_logger().info(
            f"Reached number: {result.reached_number}; status={status_name}")
        rclpy.shutdown()


def main(args=None):
    rclpy.init(args=args)
    node = CountUntilClient()
    # 先发出 Goal，再 spin 等待各个回调被依次触发
    node.send_goal()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    main()
